Azure Active Directory support

julsemaan opened the pull request:
# Description
Adds support for authenticating and authorizing Azure AD users 

# Impacts
FreeRADIUS server (adds a new module in it)
PF authentication sources

# Delete branch after merge
YES

# Checklist
(REQUIRED) - [yes, no or n/a]
- [X] Document the feature
- [X] Add acceptance tests (TestLink)

# NEWS file entries
## New Features
* Added support for Azure AD authentication

julsemaan commented:
@extrafu, I included the perl module from https://github.com/jimdigriz/freeradius-oauth2-perl which is GPL licensed.

Not sure if this should be packaged as a separate dependency or if it can stay in our repo

The file in question is: raddb/mods-config/perl/oauth2.pm

fdurand commented on docs/installation/azuread.asciidoc:
```diff
@@ -0,0 +1,83 @@
+   1. Strip on the portal
+   1. Strip on the admin
+   1. Strip in RADIUS authorization
+  1. Save the realm
```
Don't we have to set no strip in the freeradius proxy options ?

julsemaan replied:
Not according to my tests, I got the username unstripped in httpd.aaa with that configuration

fdurand commented on lib/pf/UnifiedApi/Controller/Config/Sources.pm:
```diff
@@ -27,8 +27,10 @@ use pfappserver::Form::Config::Source;
+use pfappserver::Form::Config::Source::AzureAD;
 use pfappserver::Form::Config::Source::Blackhole;
 use pfappserver::Form::Config::Source::Authorization;
+use pfappserver::Form::Config::Source::AzureAD;
```
duplicate

fdurand commented on raddb/mods-config/perl/oauth2.pm:
```diff
@@ -0,0 +1,400 @@
+
+# https://wiki.freeradius.org/modules/Rlm_perl#logging is wrong...
+use constant {
+	L_AUTH			=> 2,
```
You will have to fix the identation

fdurand reviewed (commented):
Except few details all looks to be ok. (just need to test it after merge)
@satkunas what about the admin gui part ?

julsemaan commented:
Adjusted based on the comments and rebased the branch.

The admin looks good with my rebase with @satkunas' latest changes but I'll let him be the judge on whether or not something else needs to be adjusted in the frontend

satkunas commented:
@julsemaan the ui portion looks good.

satkunas commented on html/pfappserver/root/src/views/Configuration/sources/_components/FormTypeAzureAD.vue:
```diff
@@ -0,0 +1,93 @@
+
+// @vue/component
+export default {
+  name: 'form-type-edir',
```
name: 'form-type-azure-a-d'

julsemaan commented:
@extrafu, can you take a look at the doc

We'll also need to discuss where is the best place to put this in our install guide

extrafu commented on docs/installation/azuread.asciidoc:
```diff
@@ -0,0 +1,83 @@
+
+== Azure AD integration
+
+PacketFence supports integrating with the Azure Active Directory for authenticating users on the captive-portal, the admin interface and for 802.1x users using EAP-TTLS PAP. If your only goal is to authenticate users on the captive-portal, using the OpenID implementation of Azure AD may be better suited. This section is aimed at providing username/password authentication through Azure AD.
```
everywhere: 802.1x -> 802.1X

extrafu replied:
everywhere:  captive-portal -> captive portal

extrafu commented on raddb/mods-config/perl/oauth2.pm:
```diff
@@ -0,0 +1,400 @@
+use strict;
```
We need to state (c) ownership on that file + license.

julsemaan replied:
Just to make sure, would adding the following perldoc work:

```
=head1 COPYRIGHT

Copyright Alexander Clouter (https://github.com/jimdigriz)

Licence: https://github.com/jimdigriz/freeradius-oauth2-perl/blob/master/LICENSE

=cut
```

extrafu replied:
Yes it should be OK.

julsemaan commented:
PR adjusted, I think all that is left is to decide where to put the section in our doc and we're good for merging

extrafu commented:
> PR adjusted, I think all that is left is to decide where to put the section in our doc and we're good for merging

I suggest we put it here: https://www.packetfence.org/doc/PacketFence_Installation_Guide.html#_authentication_mechanisms

after "External API Authentication"

julsemaan commented:
Doc has been placed into the install guide, should be ready for merging AFAIK

extrafu commented:
@fdurand Please merge and update the NEWS file!

fdurand commented on raddb/mods-config/perl/oauth2.pm:
```diff
@@ -0,0 +1,408 @@
+
+use strict;
+use warnings;
+
```
use lib qw(
    /usr/local/pf/lib_perl/lib/perl5
);

julsemaan replied:
Don
